modules/http/src/main/scala/higherkindness/mu/http/implicits.scala

```diff
@@ -0,0 +1,128 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package higherkindness.mu.http
+
+import cats.ApplicativeError
+import cats.effect._
+import cats.implicits._
+import fs2.{RaiseThrowable, Stream}
+import io.grpc.Status.Code._
+import org.typelevel.jawn.ParseException
+import io.circe._
+import io.circe.jawn.CirceSupportParser.facade
+import io.circe.syntax._
+import io.grpc.{Status => _, _}
+import jawnfs2._
+import org.http4s._
+import org.http4s.dsl.Http4sDsl
+import org.http4s.Status.Ok
+import scala.util.control.NoStackTrace
+
+object implicits {
+
+  implicit class MessageOps[F[_]](private val message: Message[F]) extends AnyVal {
+
+    def jsonBodyAsStream[A](
+        implicit decoder: Decoder[A],
+        F: ApplicativeError[F, Throwable]): Stream[F, A] =
+      message.body.chunks.parseJsonStream.map(_.as[A]).rethrow
+  }
+
+  implicit class RequestOps[F[_]](private val request: Request[F]) {
+
+    def asStream[A](implicit decoder: Decoder[A], F: ApplicativeError[F, Throwable]): Stream[F, A] =
+      request
+        .jsonBodyAsStream[A]
+        .adaptError { // mimic behavior of MessageOps.as[T] in handling of parsing errors
+          case ex: ParseException =>
+            MalformedMessageBodyFailure(ex.getMessage, Some(ex)) // will return 400 instead of 500
+        }
+  }
+
+  implicit class ResponseOps[F[_]](private val response: Response[F]) {
+
+    implicit def EitherDecoder[A, B](
+        implicit a: Decoder[A],
+        b: Decoder[B]): Decoder[Either[A, B]] = {
+      val l: Decoder[Either[A, B]] = a.map(Left.apply)
+      val r: Decoder[Either[A, B]] = b.map(Right.apply)
+      l or r
+    }
+
+    implicit private val throwableDecoder: Decoder[Throwable] =
+      Decoder.decodeTuple2[String, String].map {
+        case (cls, msg) =>
+          Class
+            .forName(cls)
+            .getConstructor(classOf[String])
+            .newInstance(msg)
+            .asInstanceOf[Throwable]
+      }
+
+    def asStream[A](
+        implicit decoder: Decoder[A],
+        F: ApplicativeError[F, Throwable],
+        R: RaiseThrowable[F]): Stream[F, A] =
+      if (response.status.code != Ok.code) Stream.raiseError(ResponseError(response.status))
+      else response.jsonBodyAsStream[Either[Throwable, A]].rethrow
+  }
+
+  implicit class Fs2StreamOps[F[_], A](private val stream: Stream[F, A]) {
+
+    implicit def EitherEncoder[A, B](implicit ea: Encoder[A], eb: Encoder[B]): Encoder[Either[A, B]] =
+      new Encoder[Either[A, B]] {
+        final def apply(a: Either[A, B]): Json = a match {
+          case Left(a)  => a.asJson
+          case Right(b) => b.asJson
+        }
+      }
+
+    implicit val throwableEncoder: Encoder[Throwable] = new Encoder[Throwable] {
+      def apply(ex: Throwable): Json = (ex.getClass.getName, ex.getMessage).asJson
+    }
+
+    def asJsonEither(implicit encoder: Encoder[A]): Stream[F, Json] = stream.attempt.map(_.asJson)
```

Comment:
```scala
def asJsonEither(implicit encoder: Encoder[A]): Stream[F, Json] = 
  stream.attempt.bimap(e => UnexpectedError(e.getClass.getName, Option(e.getMessage)), _.asJson)
```

Reply:
I'm sorry @fedefernandez, I don't really understand how you expect to instantiate the `UnexpectedError` without having the `Status`.

Maybe you're proposing `case class UnexpectedError(status: String, msg: Option[String] = None)`, instead of `case class UnexpectedError(status: Status, msg: Option[String] = None)`?